### PR TITLE
#317 Check if app exist before running

### LINKED ISSUE #317: fail fast in client:run if executable/app doesn't exist
This can happen if link goal wasn't called first, for instance.

### PR DESCRIPTION
**Pull request by jperedadnr**
Fixes #317

### CODE REVIEW

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=4944085&bid=16343698)

File: src/main/java/com/gluonhq/substrate/target/IosTargetConfiguration.java [hunk 1]
```diff
@@ -47,6 +47,7 @@
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Objects;
```
Comment (eugener) [inline src/main/java/com/gluonhq/substrate/target/IosTargetConfiguration.java]: Codacy Issue found: [Unused import - java.util.Objects.](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=4944085)

Comment (eugener) [inline src/main/java/com/gluonhq/substrate/target/IosTargetConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.Objects'](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=4944085)

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 2]
```diff
@@ -287,6 +287,11 @@ public String run(Path appPath, String appName) throws IOException, InterruptedE
      */
     @Override
     public boolean runUntilEnd() throws IOException, InterruptedException {
+        Path app = Objects.requireNonNull(paths.getAppPath(), "Application path can't be null")
+                .resolve(Objects.requireNonNull(projectConfiguration.getAppName(), "Application name can't be null"));
+        if (!Files.exists(app)) {
+            throw new IOException("Application not found at path " + app.toString());
+        }
         Process runProcess = startAppProcess(paths.getAppPath(), projectConfiguration.getAppName());
```
Comment (abhinayagarwal) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: Although, there is inherently nothing wrong with this code, the intent would be clearer if these tests have their own lines.

```
final Path appPath = Objects.requireNonNull(paths.getAppPath(), "Application path can't be null");
final String appName = Objects.requireNonNull(projectConfiguration.getAppName(), "Application name can't be null");
if (!Files.exists(appPath.resolve(appName))) {
    throw new IOException("Application not found at path " + appName);
}
Process runProcess = startAppProcess(appPath, appName);
```

Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: changed

File: src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java [hunk 3]
```diff
@@ -287,7 +287,14 @@ public String run(Path appPath, String appName) throws IOException, InterruptedE
      */
     @Override
     public boolean runUntilEnd() throws IOException, InterruptedException {
-        Process runProcess = startAppProcess(paths.getAppPath(), projectConfiguration.getAppName());
+        Path appPath = Objects.requireNonNull(paths.getAppPath(),
+                "Application path can't be null");
+        Path app = appPath.resolve(Objects.requireNonNull(projectConfiguration.getAppName(),
+                "Application name can't be null"));
```
Comment (abhinayagarwal) [inline src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: I particularly don't like this line: checking for `null` inside a method call. May be its just me :)

Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java]: Fixed

**Pull request merged**